### ISSUE TITLE
Question: Shared data between typescript transformers

### ISSUE THREAD

**Issue by LukasMachetanz** [nonara/ts-patch#9]
Hey everyone! I know that this is maybe not the appropriate channel for this question, but does someone know if it is possible to share data between two typescript transformers? Greets Lukas

**Comment by nonara** [comment]
Hi Lukas! It could be reasonable to use a global variable, but you can also import a shared module from multiple transformers that houses what you'd like to share.

An example would be:

```TypeScript
// shared.js
export const myData = {
  a: 1,
  b: 2
}
```

You can import from `shared.js` in the transformers and alter / amend any object that it exports, so the module can serve as a shared data-space between transformers.

**Comment by LukasMachetanz** [comment]
So it should be possible to have for example a shared array? E.g. the first transformer pushes something into the array and the second transformer is able to access it?

**Comment by nonara** [comment]
> So it should be possible to have for example a shared array? E.g. the first transformer pushes something into the array and the second transformer is able to access it?

Absolutely! 

```TypeScript
// shared.js
export const myArr = [];
```

Then in transformers `import { myArr } from './shared'` and read/push into it as needed.

When node processes a `require()` or `import`, it checks if the module has been loaded. If it hasn't, it will load it into cache. All subsequent imports get the cached module. This means that `myArr` is initialized once and any file that imports it will get a reference to the same array.

**Comment by LukasMachetanz** [comment]
Thanks for the answer. I have to try it. 

I seems that you have quite some experience with this stuff. Therefore I wanted to ask if the topic discussed here would be the best solution for the described problem here: https://stackoverflow.com/questions/61718877/typescript-transformer-follow-module-imports

I would really appreciate to hear your thoughts if you can find time to read it. :)

**Comment by nonara** [comment]
I'd be glad to help, but the example files you gave aren't discernible for me. 

## Project Structure
A project with transformer should have:

1. Project Source Files (which are fed to transformer) -> All should be included in tsconfig.json
2. A transformer file -> This is referenced in tsconfig.json under plugins: { transformer: "path/to/transformer" }

Your transformer file is _not part of the main Project_. Ideally, it should not include any imports to project files.

With that said, it is possible once your transformer is loaded to look through all source files in the project by accessing the `Program` instance, which `ts-patch` provides for you. You can use this to look for a specific function or identifier.

## Example Tranformer
You can do this by having a transformer like the following:

```TypeScript
export default function (program: Program) {
  /* Pre-transform logic (this runs before transformation starts) */
  const sourceFiles = program.getSourceFiles().filter(f => program.getRootFileNames().conains(f.fileName)); // This is an array of `SourceFile` nodes in your project
  // Logic to find whatever node / info you're looking for here

  return (ctx: ts.TransformationContext) => (node: ts.Node): ts.Node => {
    // transform logic here
  }
}
```

## Recommended Tools
I'd *highly* recommend looking at [ts-query](https://www.npmjs.com/package/@phenomnomnominal/tsquery). It can help you locate nodes on the AST with specific attributes, by name, etc. Very similar to CSS & very helpful!

You can also use [ts-query playground](https://tsquery-playground.firebaseapp.com/) to help you test matching nodes

And if you want to see what node structure looks like for a file, use:
https://ts-ast-viewer.com/

## Recommended reading
This will probably help you more than anything I've written. This is the definitive guide on writing transformers:

https://github.com/madou/typescript-transformer-handbook

Finally, if you get stuck and you have an example I can look at that fits the guidelines I mentioned above (Project Structure), feel free to post it to stackblitz and link it in this thread, and I'll have a look.

**Comment by LukasMachetanz** [comment]
Thank's for the detailed answer. But as a matter of fact I was already aware of the most mentioned points. Maybe my problem statement at stack overflow is just not good enough.

Therefore I tried to push the basic example to this repository:
https://github.com/LukasMachetanz/ttransformer-playground 

In order to run the transformer you hopefully just need this steps:
npm i
npm run transform

I tried to add comments in the source code to explain myself.

The following two screenshots should explain the comment in the file transformer.ts in more detail.

<img width="1680" alt="Screenshot 2020-05-14 at 13 40 28" src="https://user-images.githubusercontent.com/44839051/81930271-a6b43280-95e8-11ea-9fe5-5174b69fc2e8.png">
<img width="1680" alt="Screenshot 2020-05-14 at 13 40 59" src="https://user-images.githubusercontent.com/44839051/81930276-a87df600-95e8-11ea-8047-ef7e4cba4309.png">

Maybe with the source code you understand my problem describe at https://stackoverflow.com/questions/61718877/typescript-transformer-follow-module-imports better.

I am really interested what in your opinion is the best way to solve this. Maybe it's already the approach to use to transformers and share the data between them. But I have to try if this would really be possible.

I am looking forward for your answer.
Greets Lukas

**Comment by nonara** [comment]
Thanks! That clears it up. I've posted the answer on your SO question.

**Comment by LukasMachetanz** [comment]
Nice. Thank you very much. That's what I was looking for.

But know I would have a follow up question: How can I transform/adjust the found class definition?

```
const aliasedSymbol = typeChecker.getAliasedSymbol(argumentSymbol);
const classDeclaration = aliasedSymbol.declarations.find((declaration) => ts.isClassDeclaration(declaration));

if (classDeclaration && ts.isClassDeclaration(classDeclaration)) {
   // TODO: Is it possible to transform the node here?
}

```
I know how to do it the "normal" way (e.g. if I traverse the node while the normal traversation process. In this case I just have to return the transformed node as explained in https://github.com/madou/typescript-transformer-handbook#updating-a-node. But is it also possible in this case? Unfortunately I could not achieve it with my first approach. Any idea?

**Comment by nonara** [comment]
Np. If you wouldn't mind accepting the answer on SO, that will help anyone else looking.

As far as your other question, I've experimented some with directly modifying. I've had some success using the `ts.updateXNode` methods and replicating behaviour of the transformer. Overall, however, it was often unreliable and isn't really the way it's meant to be done. This can also break a lot of things if it's more than a throwaway instance of typescript, because you'd be modifying the `SourceFile` that the compiler knows about and is tracking. That breaks references, which can cause problems. It can work if you're doing like a  single TSC compile, but I wouldn't recommend it. I've moved away from doing anything like that.

What I've found to be the best route is to do your find logic *before* transformation starts. Then when it's time to transform, you'll be ready.

In your case, I would first get the symbol for your magic function, then find all calls to that function in the root files and extract the class names from the parameters. If you do all of that after the Program entry point but, *before* your transformer entry point, then transformation will be easy, because you know exactly which classes in which files need to be transformed.

I'm fairly certain that there is an easy way to get references *to* a certain symbol, but if not, you could make one rather easily.

Your best best on things like this will be to have a copy of the TS compiler source code and search it. That's how I've learned 99% of what I know so far. There's not a lot of info out there! I use WebStorm's `Find in Path` on the src directory. That with some occasional regex patterns generally helps me find what I need.

There's also a discord room #compiler - occasionally you can get answers there, and you'll learn some things just by idling. There are quite a few folks who are more knowledgeable on the compiler hanging around.
https://discord.com/invite/typescript

Good luck

**Comment by LukasMachetanz** [comment]
Done.

Again. Thanks thanks thanks. You are right; there isn't a lot of info out there. That's why I really appreciate what you are doing.

I am looking forward trying out your suggestions.

But just to be clear:

> after the program entry point, but before your transformer entry point

Are you referring to this line:
https://github.com/LukasMachetanz/ttransformer-playground/blob/250fac462218b5485e7c2b71ea4f4fa9a03734bb/transformer.ts#L59

So everything defined in the `transformerProgram` block.

And everything defined in the most inner `getTransformerFactory` block would be after my transformer entry point.

Is it like this? So would my suggestion be the correct place? If I have this information I think I am good to fiddle around with new input. :)

**Comment by nonara** [comment]
> Is it like this? So would my suggestion be the correct place? If I have this information I think I am good to fiddle around with new input. :)

Correct! 😃

**Comment by LukasMachetanz** [comment]
Okay nice. Thank you. 

Maybe one other thing I had in mind:

> In your case, I would first get the symbol for your magic function, then find all calls to that function in the root files and extract the class names from the parameters. If you do all of that after the Program entry point but, before your transformer entry point, then transformation will be easy, because you know exactly which classes in which files need to be transformed.

How would you actually do this without using the transformer? (after the Program entry point but, before your transformer entry point)

**Comment by LukasMachetanz** [comment]
Sorry to bother you again; but do you have any idea why a node maybe has no symbol? So why `typeChecker.getSymbolAtLocation(node)` does return undefined? 

````
function getVisitorToTransformTargets(program: ts.Program, context: ts.TransformationContext, transformerTargets: ts.Symbol[]): ts.Visitor {
    const typeChecker = program.getTypeChecker();

    const transformVisitor: ts.Visitor = (node: ts.Node): ts.VisitResult<ts.Node> => {
        // =========================
        // === START - Transform ===

        // ================================
        // === HANDLE CLASS DECLARATION ===
        // ================================
        if (ts.isClassDeclaration(node)) {
            console.log("CLASS DECLARATION");
            const symbol = typeChecker.getSymbolAtLocation(node);
            console.log(symbol); // TODO: WHY UNDEFINED???
        }

        // === END - Transform ===
        // =======================

        return ts.visitEachChild(node, transformVisitor, context);
    };

    return transformVisitor;
}
```

**Comment by nonara** [comment]
Try using `node.symbol`. A class should always have a `Symbol`. I'm not sure why getSymbolAtLocation doesn't work in those cases. IIRC, I've encountered that as well.

**Comment by LukasMachetanz** [comment]
Thank you so much. I can't say it often enough. You saved me again. Thanks thanks thanks. 

You are the reason why I managed to achieve the first working parts of my goal. I consider myself really lucky that I got your help. 

I know that you are not my personal advisor and that I definitely put your helpfulness to the test, but if you are still not tired of answering questions you maybe enjoy having a look on this stack overflow issue. ;)

https://stackoverflow.com/q/61989944/13513418

If this is not the case I can definitely understand it and I would like to finally thank you again. Apart from your amazing mom package your cooperativeness is amazing.
